File: internal/handlers/pd_subject.go
package handlers

import (
	"net/http"
	"net/url"
	"strconv"

	"github.com/labstack/echo/v4"

	"systemburo/internal/pdsubject"
)

// PDSubjectHandler - сведения о субъекте персональных данных (#2356): поиск человека,
// состав того, что система о нём хранит, и выгрузка справки с записью в журнал выдач.
//
// Раздел закрыт правом page.admin.pd_subject, выгрузка - парным правом
// action.pd_subject.export: увиденное на экране остаётся в системе, а выданный файл
// уходит третьему лицу и живёт дальше сам по себе.
type PDSubjectHandler struct {
	service *pdsubject.Service
}

func NewPDSubjectHandler(s *pdsubject.Service) *PDSubjectHandler {
	return &PDSubjectHandler{service: s}
}

// Find godoc
// @Summary      Поиск человека по имени
// @Description  Поиск человека по имени или по номеру документа. Склейки по имени нет: однофамильцы существуют, решает оператор. Право page.admin.pd_subject.
// @Tags         pd-subject
// @Produce      json
// @Security     BearerAuth
// @Param        fio query string false "Фамилия Имя Отчество"
// @Param        document query string false "Номер паспорта или патента"
// @Success      200 {array} pdsubject.Candidate
// @Router       /pd-subject/candidates [get]
func (h *PDSubjectHandler) Find(c echo.Context) error {
	found, err := h.service.Find(c.Request().Context(), c.QueryParam("fio"), c.QueryParam("document"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}
	return RespondSuccess(c, found)
}

// Report godoc
// @Summary      Сведения о человеке
// @Description  Что система хранит о человеке: сведения, заявки, проходы, посты. Право page.admin.pd_subject.
// @Tags         pd-subject
// @Produce      json
// @Security     BearerAuth
// @Param        registry_id query int true "Идентификатор записи реестра"
// @Success      200 {object} pdsubject.ReportResponse
// @Router       /pd-subject/report [get]
func (h *PDSubjectHandler) Report(c echo.Context) error {
	registryID, _ := strconv.Atoi(c.QueryParam("registry_id"))
	employeeID, _ := strconv.Atoi(c.QueryParam("employee_id"))
	if registryID <= 0 && employeeID <= 0 {
		return echo.NewHTTPError(http.StatusBadRequest, "укажите registry_id или employee_id")
	}
	resp, err := h.service.Report(c.Request().Context(), registryID, employeeID)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}
	return RespondSuccess(c, resp)
}

// Export godoc
// @Summary      Выгрузить справку о человеке
// @Description  Собирает справку файлом и пишет выдачу в журнал. Получатель и реквизиты запроса обязательны. Права page.admin.pd_subject и action.pd_subject.export.
// @Tags         pd-subject
// @Accept       json
// @Produce      application/octet-stream
// @Security     BearerAuth
// @Param        request body pdsubject.ExportRequest true "Кому и по какому запросу"
// @Success      200 {file} binary
// @Router       /pd-subject/export [post]
func (h *PDSubjectHandler) Export(c echo.Context) error {
	var req pdsubject.ExportRequest
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid body: "+err.Error())
	}
	if err := c.Validate(&req); err != nil {
		return err
	}

	data, name, mime, err := h.service.Export(c.Request().Context(), req,
		pdsubject.Actor{UserID: GetUserID(c), Username: GetUsername(c)})
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}
	// Имя справки кириллическое, а заголовки HTTP - ASCII: без filename*=UTF-8''
	// браузер сохраняет файл кракозябрами (поймано ручной проверкой на стенде).
	// Тот же приём, что у выгрузки бланков в attachment_blank.go: ASCII-запасное имя
	// для старых клиентов плюс закодированное настоящее.
	c.Response().Header().Set(echo.HeaderContentDisposition,
		`attachment; filename="pd-subject-report.xlsx"; filename*=UTF-8''`+url.PathEscape(name))
	return c.Blob(http.StatusOK, mime, data)
}

// Disclosures godoc
// @Summary      Журнал выдач сведений
// @Description  Кому, когда и по какому запросу выдавали сведения. Право page.admin.pd_subject.
// @Tags         pd-subject
// @Produce      json
// @Security     BearerAuth
// @Param        registry_id query int false "Только по одному человеку"
// @Success      200 {array} models.PDDisclosure
// @Router       /pd-subject/disclosures [get]
func (h *PDSubjectHandler) Disclosures(c echo.Context) error {
	registryID, _ := strconv.Atoi(c.QueryParam("registry_id"))
	employeeID, _ := strconv.Atoi(c.QueryParam("employee_id"))
	limit, _ := strconv.Atoi(c.QueryParam("limit"))
	if limit <= 0 || limit > 500 {
		limit = 100
	}
	entries, err := h.service.Disclosures(c.Request().Context(), registryID, employeeID, limit)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}
	return RespondSuccess(c, entries)
}

File: frontend/src/api/pdSubject.js
/** Сведения о субъекте персональных данных (#2356). */
import { apiRequest, apiRequestRaw } from '@/api/client';

/**
 * apiRequest отдаёт Response с подменённым json(): конверт {success,data} он
 * разворачивает уже внутри него. Поэтому ответ обязателен к чтению через
 * `await response.json()` - вернув сам Response, экран получил бы объект вместо
 * массива и молча показал «ничего не найдено».
 */
async function readJson(response) {
  if (!response.ok) {
    const body = await response.json().catch(() => null);
    throw new Error((body && body.message) || 'Запрос не выполнен');
  }
  return response.json();
}

/**
 * Поиск человека по имени или по номеру документа. По имени склейки нет - решает
 * оператор; по документу находится ровно один человек.
 */
export async function findSubjectCandidates({ fio, document }) {
  const q = document
    ? `document=${encodeURIComponent(document)}`
    : `fio=${encodeURIComponent(fio)}`;
  const response = await apiRequest(`/pd-subject/candidates?${q}`);
  return (await readJson(response)) || [];
}

/**
 * Состав сведений о человеке. Цель - запись реестра ИЛИ строка заявки: у человека,
 * попавшего в систему одной подачей, записи реестра нет вовсе.
 */
export async function fetchSubjectReport({ registryId, employeeId }) {
  const q = registryId ? `registry_id=${registryId}` : `employee_id=${employeeId}`;
  const response = await apiRequest(`/pd-subject/report?${q}`);
  return readJson(response);
}

/** Журнал выдач: весь или по одному человеку. */
export async function fetchSubjectDisclosures(target = {}) {
  const { registryId, employeeId } = target;
  let q = '';
  if (registryId) q = `?registry_id=${registryId}`;
  else if (employeeId) q = `?employee_id=${employeeId}`;
  const response = await apiRequest(`/pd-subject/disclosures${q}`);
  return (await readJson(response)) || [];
}

/**
 * Выгрузка справки файлом. Ответ - двоичный файл, а не конверт, поэтому сырой
 * запрос: имя файла берём из заголовка, его формирует сервер вместе с записью в
 * журнале выдач.
 */
export async function exportSubjectReport(payload) {
  const response = await apiRequestRaw('/pd-subject/export', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    const text = await response.text();
    throw new Error(text || 'Не удалось выгрузить справку');
  }
  // Имя файла кириллическое, поэтому сервер отдаёт его в filename*=UTF-8'' (заголовки
  // HTTP - ASCII). Читаем сначала его: простой filename="..." там запасной, ASCII-шный.
  const disposition = response.headers.get('Content-Disposition') || '';
  const utf8 = disposition.match(/filename\*=UTF-8''([^;]+)/i);
  const plain = disposition.match(/filename="([^"]+)"/);
  let filename = `Сведения_о_субъекте.${payload.format || 'xlsx'}`;
  if (utf8) filename = decodeURIComponent(utf8[1]);
  else if (plain) filename = plain[1];
  return { blob: await response.blob(), filename };
}
